data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/task/ImportTask.java

```diff
@@ -0,0 +1,431 @@
+      return ImportTargetResult.builder()
+          .namespace(namespace)
+          .tableName(table)
+          .status(ImportTargetResultStatus.VALIDATION_FAILED)
+          .errors(
+              Collections.singletonList(
+                  CoreError.DATA_LOADER_COULD_NOT_FIND_PARTITION_KEY.buildMessage()))
+          .build();
+    }
+    Optional<Key> optionalClusteringKey = Optional.empty();
+    if (!clusteringKeyNames.isEmpty()) {
+      optionalClusteringKey =
+          KeyUtils.createClusteringKeyFromSource(
+              clusteringKeyNames, dataTypeByColumnName, mutableSourceRecord);
+      if (!optionalClusteringKey.isPresent()) {
+        return ImportTargetResult.builder()
+            .namespace(namespace)
+            .tableName(table)
+            .status(ImportTargetResultStatus.VALIDATION_FAILED)
+            .errors(
+                Collections.singletonList(
+                    CoreError.DATA_LOADER_COULD_NOT_FIND_CLUSTERING_KEY.buildMessage()))
+            .build();
+      }
+    }
+
+    Optional<Result> optionalScalarDBResult;
+
+    try {
+      optionalScalarDBResult =
+          getDataRecord(
+              namespace, table, optionalPartitionKey.get(), optionalClusteringKey.orElse(null));
+    } catch (ScalarDBDaoException e) {
+      return ImportTargetResult.builder()
+          .namespace(namespace)
+          .tableName(table)
+          .status(ImportTargetResultStatus.RETRIEVAL_FAILED)
+          .errors(Collections.singletonList(e.getMessage()))
+          .build();
+    }
+    ImportTaskAction importAction =
+        optionalScalarDBResult.isPresent() ? ImportTaskAction.UPDATE : ImportTaskAction.INSERT;
+
+    if (importAction == ImportTaskAction.INSERT
+        && shouldRevalidateMissingColumns(importOptions, checkForMissingColumns)) {
+      ImportSourceRecordValidationResult validationResultForMissingColumns =
+          new ImportSourceRecordValidationResult();
+      ImportSourceRecordValidator.checkMissingColumns(
+          mutableSourceRecord, columnNames, validationResultForMissingColumns, tableMetadata);
+      if (!validationResultForMissingColumns.isValid()) {
+        return ImportTargetResult.builder()
+            .namespace(namespace)
+            .tableName(table)
+            .status(ImportTargetResultStatus.MISSING_COLUMNS)
+            .errors(
+                Collections.singletonList(
+                    CoreError.DATA_LOADER_UPSERT_INSERT_MISSING_COLUMNS.buildMessage()))
+            .build();
+      }
+    }
+
+    if (shouldFailForExistingData(importAction, importOptions)) {
+      return ImportTargetResult.builder()
+          .namespace(namespace)
+          .tableName(table)
+          .importedRecord(mutableSourceRecord)
+          .importAction(importAction)
+          .status(ImportTargetResultStatus.DATA_ALREADY_EXISTS)
+          .errors(
+              Collections.singletonList(CoreError.DATA_LOADER_DATA_ALREADY_EXISTS.buildMessage()))
```

Comment:
How about adding key information in the error message for investigation?

Reply:
Thanks for the explanation! Sounds good.

How about adding a comment at https://github.com/scalar-labs/scalardb/pull/2462/files#diff-fd06988f4ced6a0d6d0b1049c69ab07df11c372c3ccb868a78afbe4d24e6f663R827 like `// These error messages will be output with key information` ?